### PR TITLE
feat(server): add graceful shutdown function

### PR DESCRIPTION
**Pull request by maryShopline**
Add graceful shutdown to gRPC server

### CODE REVIEW

File: lib/gRpcServer.js [hunk 1]
```diff
@@ -40,26 +40,39 @@ class GRpcServer {
   }
 
   run() {
-    const server = new grpc.Server();
-
+    this.server = new grpc.Server();
```
Comment (travis-shoplineapp) [inline lib/gRpcServer.js:44]: also init in constructor?
this.server = undefined

File: lib/gRpcServer.js [hunk 2]
```diff
@@ -40,26 +40,39 @@ class GRpcServer {
...
+      grpc.ServerCredentials.createInsecure()
+    );
     log('sl-express', 'info', { message: `Starting grpc server on 0.0.0.0:${this.port}` });
-    return server.start();
+    return this.server.start();
+  }
+
+  stop() {
+    return new Promise((resolve, reject) => {
+      this.server.tryShutdown(() => {
```
Comment (travis-shoplineapp) [inline lib/gRpcServer.js]: check for this.server

File: lib/gRpcServer.js [hunk 3]
```diff
@@ -40,26 +41,43 @@ class GRpcServer {
...
+  stop() {
+    return new Promise((resolve, reject) => {
+      if (this.server) {
+        this.server.tryShutdown(() => {
+          log('sl-express', 'info', {
+            message: `Shutting down grpc server on 0.0.0.0:${this.port}`,
+          });
+        });
+      }
+      resolve();
```
Comment (Seitk) [inline lib/gRpcServer.js]: I think the `resolve` call should be inside the `tryShutdown` callback

File: lib/gRpcServer.js [hunk 4]
```diff
@@ -40,26 +41,43 @@ class GRpcServer {
...
+      grpc.ServerCredentials.createInsecure()
+    );
     log('sl-express', 'info', { message: `Starting grpc server on 0.0.0.0:${this.port}` });
-    return server.start();
+    return this.server.start();
+  }
+
+  stop() {
+    return new Promise((resolve, reject) => {
+      if (this.server) {
```
Comment (Seitk) [inline lib/gRpcServer.js]: maybe add a `setTimeout` and call `reject` for setting a max wait time for gracefully shutdown

Comment (travis-shoplineapp) [reply lib/gRpcServer.js]: should check `this.server` outside otherwise it may never resolve

File: lib/gRpcServer.js [hunk 5]
```diff
@@ -40,26 +41,62 @@ class GRpcServer {
...
+    if (!this.server) return;
+
+    return new Promise((resolve, reject) => {
+      const timeoutID = setTimeout(() => {
+        log('sl-express', 'error', {
+          message: `Timeout error while trying to stop grpc server`,
+        });
+
+        reject(new Error('Timeout error while trying to stop grpc server'));
+      }, 1000 * 5);
```
Comment (Seitk) [inline lib/gRpcServer.js]: let's set this to 60 seconds, and maybe make that a ENV

Comment (maryShopline) [reply lib/gRpcServer.js]: ok, I will add a variable called like`SHUTDOWN_GRPC_SERVER_TIMEOUT_PERIOD`?

**Review by Seitk — APPROVED** [review]

**Review by travis-shoplineapp — APPROVED** [review]

**Pull request merged**